src/tracing.c

```diff
@@ -1,43 +1,93 @@
                                  .enabled = false,
                                  .emit = noopTracerEmit};
 
+static inline const char *tracer__short_file_name(const char *fname)
+{
+    static const char top_src_dir[] = "raft/";
+    const char *p;
+
+    p = strstr(fname, top_src_dir);
+    return p != NULL ? p + strlen(top_src_dir) : fname;
+}
+
+static inline const char *tracer__trace_level_name(unsigned int level)
+{
+    static const char *levels[] = {"NONE",   "ALWAYS", "FATAL", "ERROR", "WARN",
+                                   "NOTICE", "INFO",   "DEBUG", "CALL"};
+
+    return level < ARRAY_SIZE(levels) ? levels[level] : levels[0];
+}
+
+static pid_t tracer__pid_cached;
+
+static inline void tracer__emit(const char *file,
+                                unsigned int line,
+                                const char *func,
+                                unsigned int level,
+                                const char *message)
+{
+    struct timespec ts = {0};
+    struct tm tm;
+    /* Use stack pointer instead of explicit thread id */
+    size_t sp = (size_t)&sp;
+
+    clock_gettime(CLOCK_REALTIME, &ts);
+    gmtime_r(&ts.tv_sec, &tm);
+
+    fprintf(stderr,
+            "LIBRAFT[%5.5u] %04d-%02d-%02dT%02d:%02d:%02d.%09lu "
+            "%6.6x %-7s %-20s %s:%-3i %s\n",
+            tracer__pid_cached,
+
+            tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday, tm.tm_hour, tm.tm_min,
+            tm.tm_sec, ts.tv_nsec,
+
+            (unsigned)(sp & 0xffff), tracer__trace_level_name(level), func,
+            tracer__short_file_name(file), line, message);
+}
+
 static inline void stderrTracerEmit(struct raft_tracer *t,
                                     const char *file,
-                                    int line,
+                                    unsigned int line,
+                                    const char *func,
+                                    unsigned int level,
                                     const char *message)
 {
-    (void)t;
-    struct timespec ts = {0};
-    /* ignore errors */
-    clock_gettime(CLOCK_REALTIME, &ts);
-    int64_t ns = ts.tv_sec * 1000000000 + ts.tv_nsec;
-    fprintf(stderr, "LIBRAFT   %" PRId64 " %s:%d %s\n", ns, file, line,
-            message);
+    assert(t->level < TRACE_NR);
+    if (level <= t->level)
+        tracer__emit(file, line, func, level, message);
 }
 struct raft_tracer StderrTracer = {.impl = NULL,
                                    .enabled = false,
                                    .emit = stderrTracerEmit};
 
 void raft_tracer_maybe_enable(struct raft_tracer *tracer, bool enabled)
 {
-    if (getenv(LIBRAFT_TRACE) != NULL) {
+    const char *trace_level = getenv(LIBRAFT_TRACE);
+
+    if (trace_level != NULL) {
+        tracer__pid_cached = getpid();
         tracer->enabled = enabled;
+
+        tracer->level = (unsigned)atoi(trace_level);
+        tracer->level = tracer->level < TRACE_NR ? tracer->level : TRACE_NONE;
```

Comment:
What about having `TRACE_ERROR` as default level?

Reply:
That seems like it would line up poorly with our existing use of `tracef` -- there are lots of call sites where it's not an error. Also we probably want `LIBRAFT_TRACE=1` to keep working

Reply:
That seems like it would line up poorly with our existing use of tracef -- there are lots of call sites where it's not an error. Also we probably want LIBRAFT_TRACE=1 to keep working

Reply:
If I'm not mistaken `LIBRAFT_TRACE=1` translates to `TRACE_ALWAYS`, which will still be shown if the default trace->level is `TRACE_ERROR`.

Reply:
> Also we probably want LIBRAFT_TRACE=1 to keep working

Agree, `LIBRAFT_TRACE=1` should probably mean to trace everything.

Reply:
Implication: 
```
#define Tracef(TRACER, ...) Tracef0(TRACER, TRACE_ALWAYS, __VA_ARGS__)
```

Reply:
Another suggestion is `LIBRAFT_TRACE=1` means 'LIBRAFT_TRACE=DEBUG` this is also in line with the current use of `tracef` as a very verbose logging system and would lead to
`#define Tracef(TRACER, ...) Tracef0(TRACER, TRACE_DEBUG, __VA_ARGS__)`

Reply:
Detailed tracing will be enabled when the environment variable `LIBRAFT_TRACE` is set upon startup.                                                             
The value of it can be in `[0..5]` range and reperesents a tracing level, where                                                                                 
`0` means "no traces" emitted, `5` enables minimum (FATAL records only), and `1`                                                                                
enables maximum verbosity (all: DEBUG, INFO, WARN, ERROR, FATAL records).